EIPS/eip-6190.md

```diff
@@ -0,0 +1,56 @@
+---
+eip: 6190
+title: Functional SELFDESTRUCT
+description: Changes SELFDESTRUCT to only cause a finite number of state changes
+author: Pandapip1 (@Pandapip1), Alex Beregszaszi (@axic)
+discussions-to: https://ethereum-magicians.org/t/eip-6190-functional-selfdestruct/12232
+status: Draft
+type: Standards Track
+category: Core
+created: 2022-12-20
+requires: 2929, 6189
+---
+
+## Abstract
+
+Changes `SELFDESTRUCT` to only cause a finite number of state changes.
+
+## Motivation
+
+The `SELFDESTRUCT` instruction has a fixed price, but is unbounded in storage/account changes (it needs to delete all keys). This has been an outstanding concern for some time.
+
+Furthermore, with *Verkle trees* accounts will be organised differently. Account properties, including storage, would have individual keys. It would not be possible to traverse and find all used keys. This makes `SELFDESTRUCT` very challenging to support in Verkle trees. This EIP is a step towards supporting `SELFDESTRUCT` in Verkle trees.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### `SELFDESTRUCT` Behaviour
+
+Instead of destroying the contract at the end of the transaction, instead, the following will occur at the end of the transaction in which it is invoked:
+
+1. The contract's code is set to `0x1`, and its nonce is set to `2^64-1`.
+2. Starting at the contract's address plus one, while either that address's code or nonce are not equal to zero, increment it. The contract's `0`th storage slot is set to that address.
+3. If the contract was self-destructed after the call being forwarded by one or more alias contracts, the alias contract's `0`th storage slot is set to the address calculated in step 2.
+4. The contract's balance is transferred, in its entirety, to the address on the top of the stack.
+5. The top of the stack is popped.
+
+### Gas Cost of `SELFDESTRUCT`
+
+The base gas cost of `SELFDESTRUCT` is set to `5000`. The gas cost of `SELFDESTRUCT` is increased by `5000` for each alias contract that forwarded to the contract being self-destructed. Finally, the [EIP-2929](./eip-2929.md) gas cost increase is applied.
+
+## Rationale
+
+This EIP is a step towards supporting `SELFDESTRUCT` in Verkle trees. It is a minimal change to the current behaviour.
```

Comment:
This reads more like motivation. I'd probably explain how you calculated the gas costs here, why `0x1` and `2^64-1`, and stuff like that.